generator/default.p9k

```diff
@@ -280,68 +313,202 @@ function __p9k_prompt_custom() {
+# @args
+#   $1 string The serialized segment data
 ##
-function __p9k_build_right_prompt() {
-  [[ ${P9K_RIGHT_PROMPT_ELEMENTS[1]:-} == "" ]] && return
-  local index=1
-  local element joined
-  for element in "${P9K_RIGHT_PROMPT_ELEMENTS[@]}"; do
-    # Check if segment should be joined
-    [[ "${element[-7,-1]}" == '_joined' ]] && joined=true || joined=false
+function __p9k_refresh_cache_item() {
+  local -a segment_meta=("${(@s:·|·:)1}")
 
-    # Remove joined information in direct calls
-    element="${element%_joined}"
+  __p9k_segment_cache["${segment_meta[2]}::${segment_meta[3]}"]="${1}"
+}
 
-    # Check if it is a custom command, otherwise interpet it as a prompt.
-    if [[ $element[0,7] =~ "custom_" ]]; then
-      "__p9k_prompt_custom" "right" "$index" ${joined} $element[8,-1]
-    else
-      "prompt_$element" "right" "$index" $joined
-    fi
+###############################################################
+# @description
+#   Exchange result of prompt_<segment> function in the cache and
+#   trigger re-rendering of prompt.
+##
+# @args
+#   # $1 job name, e.g. the function passed to async_job
+#   $2 return code
+#   $3 resulting (stdout) output from job execution
+#   $4 execution time, floating point e.g. 0.0076138973 seconds
+#   $5 resulting (stderr) error output from job execution
+#   $6 has next result in buffer (0 = buffer empty, 1 = yes)
+##
+function __p9k_async_callback() {
+  local JOB="${1}" CODE="${2}" RAW_SEGMENT_DATA="${3}" EXEC_TIME="${4}" ERR="${5}" NEXT_BUFFER="${6}"
+
+  # ignore the async evals used to alter worker environment
+  if [[ "${JOB}" == "[async/eval]" ]] || \
+     [[ "${JOB}" == ";" ]] || \
+     [[ "${JOB}" == "[async]" ]]; then
+    return
+  fi
 
-    index=$((index + 1))
+  # Exit early, if ${RAW_SEGMENT_DATA} is empty
+  if [[ -z "${RAW_SEGMENT_DATA}" ]]; then
+    return
+  fi
+
+  # split ${RAW_SEGMENT_DATA} into an array - see https://unix.stackexchange.com/a/28873
+  local segment_meta=("${(@s:·|·:)RAW_SEGMENT_DATA}") # split on delimiter "·|·" (@s:<delim>:)
+  local cache_key="${segment_meta[2]}::${segment_meta[3]}"
+  __p9k_segment_cache["${cache_key}"]="${RAW_SEGMENT_DATA}"
+
+  # Trigger re-rendering if we do not wait for other jobs
+  [[ "${NEXT_BUFFER}" == "0" ]] && __p9k_render "true"
+}
+
+###############################################################
+# @description
+#   P9K Render function.
+#   Goes through cache and renders each entry.
+##
+# @args
+#   $1 - boolean True if rendered through ZLE widget (in async mode)
+##
+function __p9k_render() {
+  # Resets
+  CURRENT_BG='NONE'
+  CURRENT_RIGHT_BG='NONE'
+  PROMPT=''
+  RPROMPT=''
+  # __p9k_unsafe must be a global variable, because we set
+  # PROMPT='$__p9k_unsafe[left]', so without letting ZSH
+  # expand this value (single quotes). This is a workaround
+  # to avoid double expansion of the contents of the PROMPT.
+  typeset -gAh __p9k_unsafe=()
+
+  # Process Cache
+  for data in "${(Oa@v)__p9k_segment_cache}"; do
```

Comment:
Another finding.

The use of expansion flag `(Oa@v)` couldn't ensure an iteration based on the array index.

Here's what I find with setting many segments on the left prompt.

```zsh
# example conf
# 12 segments on the left, and 0 on the right
P9K_LEFT_PROMPT_ELEMENTS=(context dir vcs::async node_version::async nvm::async rbenv::async anaconda::async pyenv::async status command_execution_time background_jobs time)

P9K_RIGHT_PROMPT_ELEMENTS=()
```

```zsh 
# result
for data in "${(Oa@v)__p9k_segment_cache}"; do echo "$data"; done
CONTEXT_DEFAULT·|·left·|·1·|·false·|··|··|·false
DIR_HOME_SUBFOLDER·|·left·|·2·|·false·|·…/dritter---powerlevel9k·|··|·true
VCS_CLEAN·|·left·|·3·|·false·|· conditional_async·|· ·|·true
NODE_VERSION·|·left·|·4·|·false·|··|··|·false
NVM·|·left·|·5·|·false·|··|··|·false
COMMAND_EXECUTION_TIME·|·left·|·10·|·false·|··|··|·false
RBENV·|·left·|·6·|·false·|·2.6.3·|··|·false
BACKGROUND_JOBS·|·left·|·11·|·false·|·0·|··|·false
ANACONDA·|·left·|·7·|·false·|·()·|··|·false
TIME·|·left·|·12·|·false·|·%D{%H:%M:%S}·|··|·true
PYENV·|·left·|·8·|·false·|·3.6.8:2.7.16·|··|·false
STATUS_OK·|·left·|·9·|·false·|··|· ·|·true
```

The sequence of `BACKGROUND_JOBS` is not based on its index from array `P9K_LEFT_PROMPT_ELEMENTS`.

Reply:
A pr is opened for this problem in your fork. https://github.com/dritter/powerlevel9k/pull/41